week-1-ts-ecosystem/w01e03-object-oriented-structure/src/controllers/product.ts

```diff
@@ -0,0 +1,17 @@
+import { ProductModel, ProductType } from "../models/product";
+
+export class Product<TYPE extends ProductType> {
```

Comment:
Ok. tutaj trafiamy na generyka klasy który korzysta z inteface żeby mieć odpowiedni typ.

W konsekwencji Twoja logika "kształtu" produktu - czyli `class` jest rozdzielona z `interface`
I tutaj zaczyna nam się to komplikować bo zobacz:

Z definicji klasa - to jednocześnie dane + ich kształt (Czyli typ danych).
Wiec, po co pisać podwójnie? Tutaj coś jest nadmiarowe i myślę ze to czułeś jak to pisałeś :) 

Co Ty na to żeby Koszyk `Cart` był generyczny, ale przyjmował klasy konkrentych produktów.
W tym układzie masz 3 różne klasy reprezentujące różnego rodzaju produkty.
Ponieważ nie chcesz powtarzać tych zapisów 3 razy, robisz jedną bazową klasę z elementami wspólnymi.
Natomiast dodatkowe klasy mają jeszcze `price`

I teraz masz benefit.
Owszem powiesz mi:

> Chwila, Jabi - teraz zamiast 2 plików, mam 4 robiące to samo

To prawda, ale celowo je masz bo każda z klas będzie reprezentowała model. Przykładowo:

```
models
  | - BuyNowItem.ts
  | - FreeItem.ts
  | - AuctionItem.ts
```

Wtedy widzisz jakiego rodzaju `*Item` Twój koszyk obsługuje.
A jednocześnie nie podpisujesz na siebie :
_Klauzuli ręcznego synchronizowania interface z klasą bazową_

Więc celowo się tutaj powtarzamy, ale unikami sytuacji "błędnogennej".

Daj znaka czy rozumiesz mój koncept - w razie czego wyjaśnię coś szerzej.

Reply:
Tak, tak rozumiem koncept. Generalnie w pewnym momencie się zakręciłem z tym zadaniem, ale to też z tego względu, że moje doświadczenie z kodem obiektowym jest niestety bardzo małe.
Faktycznie można wykorzystać klasę jako `kształt` danych 🤔 
Naniosłem poprawki według Twoich uwag. Jakbyś widział jeszcze jakiś potencjał na refactor to śmiało 👍